nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix

```diff
@@ -25,6 +25,8 @@ let
     inherit (cfg) consoleMode;
 
     inherit (efi) efiSysMountPoint canTouchEfiVariables;
+
+    memtest86 = if cfg.memtest86.enable then pkgs.memtest86-efi else "";
```

Comment:
If someone try to enable it anyway with `allowUnfree = false` it will bring `memtest86-efi` into scope but the next nixos-rebuild should fail to eval with a clear enough error message. So it works as intended.